solutions/rust/05-echo/code/src/resp.rs

```diff
@@ -0,0 +1,209 @@
+use anyhow::{Error, Result};
+use bytes::BytesMut;
+use tokio::{io::AsyncReadExt, io::AsyncWriteExt, net::TcpStream};
+
+const CARRIAGE_RETURN: u8 = '\r' as u8;
+const NEWLINE: u8 = '\n' as u8;
+
+#[derive(Eq, PartialEq, Clone, Debug)]
+pub enum Value {
+    /// For Simple Strings the first byte of the reply is "+".
+    String(String),
+    /// For Errors the first byte of the reply is "-".
+    Error(String),
+    /// For Bulk Strings the first byte of the reply is "$".
+    Bulk(String),
+    /// For Arrays the first byte of the reply is "*".
+    Array(Vec<Value>),
+}
+
+impl Value {
+    pub fn to_command(&self) -> Result<(String, Vec<Value>)> {
+        match self {
+            Value::Array(items) => {
+                return Ok((
+                    items.first().unwrap().unwrap_bulk(),
+                    items.clone().into_iter().skip(1).collect(),
+                ));
+            }
+            _ => Err(Error::msg("not an array")),
+        }
+    }
+
+    fn unwrap_bulk(&self) -> String {
+        match self {
+            Value::Bulk(str) => str.clone(),
+            _ => panic!("not a bulk string"),
+        }
+    }
+
+    pub fn encode(self) -> String {
+        match &self {
+            Value::String(s) => format!("+{}\r\n", s.as_str()),
+            Value::Error(msg) => format!("-{}\r\n", msg.as_str()),
+            Value::Bulk(s) => format!("${}\r\n{}\r\n", s.chars().count(), s),
+            // The other cases are not required
+            _ => panic!("value encode not implemented for: {:?}", self),
+        }
+    }
+}
+
+pub struct RespConnection {
+    stream: TcpStream,
+    buffer: BytesMut,
+}
+
+impl RespConnection {
+    pub fn new(stream: TcpStream) -> Self {
+        return RespConnection {
+            stream,
+            buffer: BytesMut::with_capacity(512),
+        };
+    }
+
+    pub async fn read_value(&mut self) -> Result<Option<Value>> {
+        loop {
+            let bytes_read = self.stream.read_buf(&mut self.buffer).await?;
+
+            if bytes_read == 0 {
+                if self.buffer.is_empty() {
+                    return Ok(None);
+                } else {
+                    return Err(Error::msg("connection closed unexpectedly"));
```

Comment:
@ThetaSinner I'm a bit confused here - Why is an empty buffer a "success" case & a non-empty buffer a failure case?